### PR TITLE
feat: Implement the Web Share Target API

### LINKED ISSUE #3214: feat: Implement the Web Share Target API
The Web Share Target API allows Excalidraw to appear in the operating system's share sheet, so when you have an `.excalidraw` file (for example in an Android file manager), you can share it to Excalidraw. The [background article](https://web.dev/web-share-target/) has all the details.

### PR DESCRIPTION
**Pull request by h7y**
## Closes
fix #3214 
fix https://github.com/excalidraw/excalidraw/issues/3157

### CODE REVIEW

File: public/manifest.json [hunk 1]
```diff
@@ -26,5 +26,18 @@
       }
     }
   ],
-  "capture_links": "new_client"
+  "capture_links": "new_client",
+  "share_target": {
+    "action": "/",
```
Comment (tomayac) [inline public/manifest.json:31]:
```suggestion
    "action": "/web-share-target",
```

We mint a fake URL here that only serves for this purpose.

File: public/manifest.json [hunk 2]
```diff
@@ -26,5 +26,18 @@
...
+    "method": "POST",
+    "enctype": "multipart/form-data",
+    "params": {
+      "files": [
+        {
+          "name": "file",
+          "accept": ["application/json", ".excalidraw"]
```
Comment (tomayac) [inline public/manifest.json:38]:
```suggestion
          "accept": [".excalidraw"]
```

Let’s be very conservative here and strictly only accept the file extension.

File: src/service-worker.js [hunk 3]
```diff
@@ -47,3 +47,24 @@ workbox.routing.registerRoute(
     plugins: [new workbox.expiration.Plugin({ maxEntries: 10 })],
   }),
 );
+
+self.addEventListener("activate", () => {
+  self.clients.claim();
+});
+
+self.addEventListener("fetch", (event) => {
```
Comment (tomayac) [inline src/service-worker.js:55]: I’m not entirely sure this is needed. I believe Workbox does this automatically, but not sure about the configuration in create-react-app. Please verify.

File: src/service-worker.js [hunk 4]
```diff
@@ -47,3 +47,24 @@ workbox.routing.registerRoute(
...
+  if (event.request.method !== "POST") {
```
Comment (tomayac) [inline src/service-worker.js:56]: This is headed in the right direction, but you want to be sure to only capture requests targeted at the /web-share-target fake URL. It’s probably easiest if you adapt my minimal example flow in this demo app: https://github.com/tomayac/fugu-greetings/search?q=share+target&type=.

File: public/manifest.json (hunk 1 above)
Comment (h7y) [reply public/manifest.json:31]: Got it!

File: src/service-worker.js (hunk 3 above)
Comment (h7y) [reply src/service-worker.js:55]: Are you referring to `  self.clients.claim();`?

Comment (tomayac) [reply src/service-worker.js:55]: Yes.

Comment (h7y) [reply src/service-worker.js:51]: https://github.com/excalidraw/excalidraw/blob/00ac804dd867d214c455292397206233fc784157/src/service-worker.js#L32

True, it's not needed!

File: public/manifest.json (hunk 2 above)
Comment (h7y) [reply public/manifest.json]: I was testing it on my android phone with Chrome 89, it's not showing Excalidraw in share sheet when trying to share `.excalidraw` files. It works with JSON. Weird!

File: src/service-worker.js (hunk 3 above)
Comment (tomayac) [reply src/service-worker.js:51]: Perfect. Thanks for checking. I was only 75% sure.

File: public/manifest.json (hunk 2 above)
Comment (tomayac) [reply public/manifest.json]: We were trying to mint our own media type, but it didn’t work with Safari. For now triggering for all .json files should work, but in the future I should investigate again if we can use something more specific.

Comment (h7y) [reply public/manifest.json:38]: Got it! I added `application/json` back 👍🏽

Comment (tomayac) [reply public/manifest.json:38]: Let’s add `application/vnd.excalidraw+json`, too, this is what we were trying to use. It’s also in the manifest a couple of lines up.

File: src/service-worker.js [hunk 5]
```diff
@@ -47,3 +47,20 @@ workbox.routing.registerRoute(
     plugins: [new workbox.expiration.Plugin({ maxEntries: 10 })],
   }),
 );
+
+self.addEventListener("fetch", (event) => {
+  if (event.request.url.endsWith("/web-share-target")) {
```
Comment (tomayac) [inline src/service-worker.js:52]: Still keep the check if the method is POST.

File: public/manifest.json (hunk 2 above)
Comment (h7y) [reply public/manifest.json]: Done 👍🏽

**Comment by tomayac** [comment]
Can you change the [default name](https://github.com/excalidraw/excalidraw/blob/47c26cd4cfb850e2627fb87635e814dd4837f264/src/appState.ts#L54) so that it ends in `".excalidraw"`? This should make sure that on Android we save with the correct extension, too.

File: src/components/App.tsx [hunk 6]
```diff
@@ -747,6 +747,10 @@ class App extends React.Component<ExcalidrawProps, AppState> {
         this.initializeScene();
       });
     }
+
+    if (window.location.search.includes("share-target")) {
```
Comment (tomayac) [inline src/components/App.tsx:751]: Let’s use the proper API here: https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams/has

File: src/components/App.tsx [hunk 7]
```diff
@@ -1277,6 +1281,20 @@ class App extends React.Component<ExcalidrawProps, AppState> {
     this.setState({ toastMessage: null });
   };
 
+  restoreFileFromShare = async () => {
+    const keys = await caches.keys();
+    const mediaCache = await caches.open(
```
Comment (tomayac) [inline src/components/App.tsx:1286]: Just use `"web-share-target"` as the cache name. The filtering is not needed (it was needed in the other implementation).

File: src/components/App.tsx [hunk 8]
```diff
@@ -1277,6 +1281,20 @@ class App extends React.Component<ExcalidrawProps, AppState> {
     this.setState({ toastMessage: null });
   };
 
+  restoreFileFromShare = async () => {
```
Comment (tomayac) [inline src/components/App.tsx:1284]: Can you wrap this in a `try … catch` block just in case?

File: src/service-worker.js [hunk 9]
```diff
@@ -47,3 +47,23 @@ workbox.routing.registerRoute(
...
+    event.request.method === "POST" &&
+    event.request.url.endsWith("/web-share-target")
+  ) {
+    return event.respondWith(
+      (async () => {
+        const formData = await event.request.formData();
+        const file = formData.get("file");
+        const keys = await caches.keys();
+        const mediaCache = await caches.open(
+          keys.filter((key) => key.startsWith("media"))[0],
```
Comment (tomayac) [inline src/service-worker.js:62]: Same as before, let’s just use `"web-share-target"` as the cache name.

File: src/service-worker.js [hunk 10]
```diff
@@ -47,3 +47,23 @@ workbox.routing.registerRoute(
...
+        );
+        await mediaCache.put("shared-file", new Response(file));
+        return Response.redirect("./?web-share-target", 303);
```
Comment (tomayac) [inline src/service-worker.js:65]:
```suggestion
        return Response.redirect("/?web-share-target", 303);
```

**Comment by tomayac** [comment]
Can you change the [default name](https://github.com/excalidraw/excalidraw/blob/47c26cd4cfb850e2627fb87635e814dd4837f264/src/appState.ts#L54) so that it ends in `".excalidraw"`? This should make sure that on Android we save with the correct extension, too.

File: src/components/App.tsx [hunk 11]
```diff
@@ -1277,6 +1281,20 @@ class App extends React.Component<ExcalidrawProps, AppState> {
...
+    const keys = await caches.keys();
+    const mediaCache = await caches.open(
+      keys.filter((key) => key.startsWith("media"))[0],
+    );
+
+    const file = await mediaCache.match("shared-file");
+    if (file) {
+      const blob = await file.blob();
+      this.loadFileToCanvas(blob);
+      await mediaCache.delete("shared-file");
```
Comment (tomayac) [inline src/components/App.tsx:1294]: At this stage we also need to remove the URL parameter from the query. The History API can do this: https://stackoverflow.com/a/41061471.

**Comment by h7y** [comment]
> Can you change the [default name](https://github.com/excalidraw/excalidraw/blob/47c26cd4cfb850e2627fb87635e814dd4837f264/src/appState.ts#L54) so that it ends in `".excalidraw"`? This should make sure that on Android we save with the correct extension, too.

On Android, this saves the file as `{filename}.excalidraw.json`.

File: src/appState.ts [hunk 12]
```diff
@@ -51,7 +51,7 @@ export const getDefaultAppState = (): Omit<
     isRotating: false,
     lastPointerDownWith: "mouse",
     multiElement: null,
-    name: `${t("labels.untitled")}-${getDateTime()}`,
+    name: `${t("labels.untitled")}-${getDateTime()}.excalidraw`,
```
Comment (tomayac) [inline src/appState.ts:54]:
```suggestion
    name: `${t("labels.untitled")}-${getDateTime()}`,
```

**Comment by tomayac** [comment]
Thanks a ton, this looks really solid now. I fear we won’t find a solution to saving/sharing with the `.excalidraw` extension in this PR (but thanks for your help with trying). I’ll do some more experiments later tonight, and then I guess we’re good to merge.

**Comment by h7y** [comment]
Awesome! This was so much fun to work fun. You are great! I learnt a lot :) Thank you so much. I'll research on getting .excalidraw to work too.

File: src/components/App.tsx [hunk 13]
```diff
@@ -747,6 +747,11 @@ class App extends React.Component<ExcalidrawProps, AppState> {
         this.initializeScene();
       });
     }
+
+    const searchParams = new URLSearchParams(window.location.search.slice(1));
+    if (searchParams.has("web-share-target")) {
+      this.restoreFileFromShare();
+    }
```
Comment (dwelle) [inline src/components/App.tsx:754]: shouldn't this override the `initializeScene()` above instead of happen in addition?

Comment (h7y) [reply src/components/App.tsx:754]: I'll do that 👍

Comment (h7y) [reply src/components/App.tsx]: Done :)

**Review by tomayac — APPROVED** [review]
LGTM

Saving now uses the `.excalidraw` extension universally and Excalidraw appears as a Web Share Target for this extension.

**Review by lipis — APPROVED** [review]

**Review by dwelle — APPROVED** [review]

**Pull request merged**